### PR TITLE
add redactList for PaymentAddress (Part 1)

### LINKED ISSUE #648: redact full shipping address from event until payment response
Redact:
 * name
 * phone
 * address line(s) 
 * part of the zip. (+4 in US) ... um, we need to work that out.

### PR DESCRIPTION
**Pull request by marcoscaceres**
closes w3c/payment-request#648 w3c/payment-request#643 w3c/payment-request#676 w3c/payment-request#692

The following tasks have been completed:

 * [X] [MDN Docs added](https://developer.mozilla.org/en-US/docs/Web/Events/shippingaddresschange) 

Implementation commitment:

 * [x] Safari (implemented in ApplePay - redacts)
 * [ ] [Chrome](https://crbug.com/821771) - planning.
 * [x] [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1435155)  - redacts.
 * Edge - keeping current behavior (no redact)

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/payment-request/pull/654.html" title="Last updated on Mar 15, 2018, 3:46 AM GMT (e21719c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/payment-request/654/f34f6b7...e21719c.html" title="Last updated on Mar 15, 2018, 3:46 AM GMT (e21719c)">Diff</a>

### CODE REVIEW

**Comment by marcoscaceres** [comment]
@domenic, @rsolomakhin, all, please note that this is incomplete, just seeking feedback on approach at this point (you will spot a "To Be Written" algo). 

I need to read E.164 spec in details, so to understand a bit more about phone number formats and structure. In any case, let me know what you think!

Renaming `PaymentAddress` might be "nice" too... but maybe too late, unless we go down the `NamedConstructor` route 🙀, which I don't think we want.. or we can just live with the crappy name.

File: index.html
Comment (rsolomakhin) [inline index.html]: Pardon my noob question. Why do we need both "constructor" and "creating" sections?

**Review by rsolomakhin — COMMENTED** [review]
Overall looks good!

File: index.html
Comment (marcoscaceres) [reply index.html]: Yeah, we might need to merge them... one is supposed to impose order when constructing from user input (should rename this section more clearly), while the other is for calling the constructor from script.

Comment (marcoscaceres) [reply index.html]: *constructing

File: index.html
Comment (domenic) [inline index.html]: The capitalization is inconsistent on these components of an address. I think consistently lowercase would be best.

File: index.html
Comment (domenic) [inline index.html]: Does exposing this in workers make sense?

File: index.html
Comment (domenic) [inline index.html]: The order here doesn't match the above, which is pretty confusing.

File: index.html
Comment (domenic) [inline index.html]: Typo "PymentAddress"

File: index.html
Comment (domenic) [inline index.html]: Do you want to distinguish between null and empty string? Null feels a bit more natural for "unset" fields.

I have a feeling we discussed this before... apologies if so.

File: index.html
Comment (domenic) [inline index.html]: You don't want to freeze the incoming dictionary. I think instead you want to set address.[[addressLine]] to a new frozen array created from details["addressLine"]. (In place of both of the above steps.)

File: index.html
Comment (domenic) [inline index.html]: What is "trimming"? :) I think you want https://infra.spec.whatwg.org/#strip-leading-and-trailing-ascii-whitespace

File: index.html
Comment (domenic) [inline index.html]: Presumably only throw the RangeError if it is not a structurally valid phone number. As phrased, this always throws it.

File: index.html
Comment (domenic) [inline index.html]: canonicalizing the phone number

File: index.html
Comment (domenic) [inline index.html]: I would phrase this as "stored in its canonical uppercase form"

File: index.html
Comment (domenic) [inline index.html]: Link to https://infra.spec.whatwg.org/#strip-and-collapse-ascii-whitespace

File: index.html
Comment (domenic) [inline index.html]: If **_tel_'s** length is greater **than** 15...

**Review by domenic — CHANGES_REQUESTED** [review]
Found some nits!

File: index.html
Comment (marcoscaceres) [reply index.html]: I think we want MAY want expose it for Payment Hander. I can redact it until we are 100% on the API.

File: index.html
Comment (marcoscaceres) [reply index.html]: We indeed have... I wanted `null` too :) But then I started playing around with integrating with `<form>` and the empty string turned out to be a better fit.

File: index.html
Comment (marcoscaceres) [reply index.html]: ah, yeah. Makes sense.

File: index.html
Comment (marcoscaceres) [reply index.html]: I'm going to change this algo to use an `AddressInit` dictionary instead. The algorithm will also take an optional "excludeList", which will exclude dictionary members as per #648. That should solve for `onpaymentaddresschange` privacy issues.

**Comment by marcoscaceres** [comment]
Commit I just pushed is still WIP... not ready for review.

**Comment by domenic** [comment]
@marcoscaceres would you like re-review of this?

**Comment by marcoscaceres** [comment]
Not yet. I will try to finish it this week. It’s still missing some stuff around phone numbers.

File: index.html
Comment (stpeter) [inline index.html:3864]: If the input form allows a wider range of UTF-8 encoded Unicode characters (not just the ASCII range), then stripping only ASCII whitespace will not necessarily yield the hoped-for result here because a fair number of Unicode code points have a Unicode general category of "Zs". For example, U+1680 (OGHAM SPACE MARK) would not get stripped unless non-ASCII space gets converted to ASCII space.

Comment (marcoscaceres) [reply index.html:3864]: yeah, that's a good point... specially because this might be coming from free form input from a end-user (e.g., "region"). 

@domenic, seems JS `String.prototpye.trim()` does mostly the right thing (some bugginess across browsers). Do you have any suggestions?

Comment (marcoscaceres) [reply index.html:3864]: Here is a test string:
```
var whiteSpace = 
"\u0009\u000A\u000B\u000C\u000D\u0020\u0085\u00A0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u2028\u2029\u202F\u205F\u3000"
whiteSpace.trim().length; // 1, because u0085 is apparently not WS. 
```

Comment (domenic) [reply index.html:3864]: We never trim Unicode whitespace on the platform. Only ASCII whitespace is considered as "allowed sloppiness" in argument values.

An alternative is to not remove such whitespace at all.

Comment (stpeter) [reply index.html:3864]: @domenic Hmm. Do we assume that non-ASCII Unicode code points with a General Category of "Zs" will be filtered out somehow? (For example, by conversion of such code points to SPACE [U+0020]?) If this is guidance that the i18n WG has provided in the past, perhaps I've just missed the pointer. I can ask in the next i18n WG call if needed.

Comment (domenic) [reply index.html:3864]: Generally if a string contains Unicode whitespace, we don't try to second-guess the user or developer, and just pass it through.

Comment (stpeter) [reply index.html:3864]: Thanks for the clarification! I just asked about it on the www-international list, too: https://lists.w3.org/Archives/Public/www-international/2018JanMar/0042.html

Comment (marcoscaceres) [reply index.html:3864]: I'm ok with doing the minimal ASCII whitespace stripping, for platform consistency. Developers can then call `.trim()` to strip out the rest if so desired.

**Comment by marcoscaceres** [comment]
So, from here, we will only take the optional privacy preserving list.

**Comment by marcoscaceres** [comment]
@domenic, this is ready to review again. 

Based on discussions with Editors, this PR primarily:

 * adds `regionCode` attribute to PaymentAddress (closes w3c/contact-picker#65)
 * introduces `excludeList` to match Safari's privacy-preserving behavior (closes w3c/payment-request#648) 
 * adds a "internal constructor" (closes w3c/payment-request#643) ... purpose being that we will eventually make that constructor public, if we do w3c/payment-request#653 (folks from Apple requested this, and we will probably support that in Firefox too).

**Comment by marcoscaceres** [comment]
Noting for those following at home: `excludeList` is optional as Chrome and Edge wish to preserve their current behavior. Firefox will make use of the `excludeList` and behave similarly to Safari. We might tweak the `excludeList` values if we find it's overly restrictive.

**Comment by rsolomakhin** [comment]
Following at home

**Comment by marcoscaceres** [comment]
Added tests https://github.com/w3c/web-platform-tests/pull/9706

**Comment by marcoscaceres** [comment]
Added implementation commitment from Firefox.

**Comment by marcoscaceres** [comment]
Actually, I want to land this in two parts. 

 1. The editorial rewrite of Physical Addresses. 
 1. Then the region code. 

That way, 1 can land now... and then 2 can land when we get another implementer to commit.

**Comment by marcoscaceres** [comment]
With just the `excludeList`, this matches WebKit and Firefox - and let's Chrome and Edge keep returning full address information for the `onpaymentaddresschange` event.

**Comment by marcoscaceres** [comment]
@domenic, when you can spare a few cycles - would like to get this approved as other PRs depend on it.

File: index.html
Comment (domenic) [inline index.html]: This is only called from one place, right? I guess it's a separate algorithm because you plan to add a constructor soon? If so make sure to call it out in the commit message, and perhaps in the spec (although that spec call out would disappear shortly).

File: index.html
Comment (domenic) [inline index.html]: Missing `<li>`

File: index.html
Comment (domenic) [inline index.html]: Missing space after `</a>`

File: index.html
Comment (domenic) [inline index.html]: Missing space and extra "set" after `</a>`

File: index.html
Comment (domenic) [inline index.html]: What constructor?

Is details meant to be an optional argument to this algorithm? If so it should be stated ("with an optional AddressInit details")

These days we're not saying "return and terminate this algorithm"; just returning suffices (like in any reasonable programming language).

File: index.html
Comment (domenic) [inline index.html]: Should we strip whitespace from country too?

File: index.html
Comment (domenic) [inline index.html]: Should we strip whitespace from each address line perhaps?

File: index.html
Comment (domenic) [inline index.html]: Missing a "represents"

File: index.html
Comment (domenic) [inline index.html]: Links to the wrong place, here and in several other "represents". You can tell because in the preview they are in code font, but they should not be.

File: index.html
Comment (domenic) [inline index.html]: a a -> a

File: index.html
Comment (domenic) [inline index.html]: as sequence -> as a sequence

File: index.html
Comment (domenic) [inline index.html]: Just say "be an AddressInit dictionary with no members present". It's technically not an object since we're not in JS-land, we're in Web IDL-land.

File: index.html
Comment (domenic) [inline index.html]: This should not be a nested list, since none of the other steps are

File: index.html
Comment (domenic) [inline index.html]: Do you have to append all of them together, or can you pick a subset? If all together, I'd say "optionally, set redactList to « ... »" If a subset, then I'd say "optionally add any of the following to redactList: ..." (the latter with no «»s)

In other places with optional steps, we include a nice note explaining why someone would do this. Can you do the same here?

File: index.html
Comment (domenic) [inline index.html]: Don't forget to pass redactList!

**Review by domenic — CHANGES_REQUESTED** [review]
Lots of editorial issues, and a few questions. Overall pretty solid stuff though!

File: index.html
Comment (marcoscaceres) [reply index.html]: As the algorithm is specified, the `redactList` is always required (even if empty). Thus passing in the empty one should be ok. 

I did this because I didn't want to write "if `redactList` was passed and if includes 'xxx' ..." over and over. Hopefully that's ok 🤞

**Comment by marcoscaceres** [comment]
@hsivonen, https://github.com/w3c/payment-request/pull/654/commits/9dd14c343e34e64b8ca98698a8ccda0736657efb should address w3c/payment-request#692.

**Comment by marcoscaceres** [comment]
I think I've addressed all the feedback 🤞 - but folks please check this carefully. Thanks again @domenic for the eagle-eyes review! 🦅

**Comment by marcoscaceres** [comment]
@domentic, when you can, waiting on your approval.

**Comment by marcoscaceres** [comment]
Added a [note to MDN](https://developer.mozilla.org/en-US/docs/Web/Events/shippingaddresschange) about some UAs redacting address information for privacy reasons.

File: index.html
Comment (domenic) [inline index.html]: Section titles seem to use "Sentence case" so far, not "Title Case"

File: index.html
Comment (domenic) [inline index.html]: Lowercase "constructor"

File: index.html
Comment (domenic) [inline index.html]: Uppercase "s"

File: index.html
Comment (domenic) [inline index.html]: Missing period

File: index.html
Comment (domenic) [inline index.html]: Misspelled "empty"

File: index.html
Comment (domenic) [inline index.html]: Move "or the empty string" to the end of the sentence.

File: index.html
Comment (domenic) [inline index.html]: This should just be a list, not a frozen array, since it's the dictionary.

File: index.html
Comment (domenic) [inline index.html]: I still can't tell what this step is saying. Is it saying redactList must be either the empty list or this list? If so, please say `Optionally, set redactList to « "organization", "phone", "recipient", "addressLine" »`.

Or is it saying that a list like just « "phone" » is valid? If so, please say `Optionally append any or all of the following to redactList: "organization", "phone", "recipient", "addressLine"`

**Review by domenic — CHANGES_REQUESTED** [review]
Sorry, a crucial step still seems unclear to me (determining redactList).

File: index.html
Comment (marcoscaceres) [reply index.html]: > I still can't tell what this step is saying. Is it saying redactList must be either the empty list or this list? If so, please say Optionally, set redactList to « "organization", "phone", "recipient", "addressLine" ».

Yep, this one. Sorry for not being clear.

File: index.html
Comment (rsolomakhin) [inline index.html]: s/is cannot/cannot

**Comment by marcoscaceres** [comment]
Thanks @rsolomakhin for also having a look! Oki, should hopefully be all good now 🤞

**Review by domenic — APPROVED** [review]

**Comment by marcoscaceres** [comment]
Thanks again @domenic!!! (the above is @rsolomakhin's fault, he started the whole gifs thing 😁)

**Comment by rsolomakhin** [comment]
Well done, sir.

**Pull request merged**
